frontend/src/components/Mermaid.js

```diff
@@ -35,13 +36,33 @@ class MermaidChart extends React.Component {
     window.callback = (e) => this.props.viewLayerFunc(e);
   }
 
+  /** @param {MouseEvent} e  */
+  onCollapseButtonClick(e) {
+    const data = e.target?.dataset?.["key"];
+    if (data == null) {
+      return;
+    }
+    this.props.toggleCollapseLayerFunc(data);
+
+    // don't fire click event on node itself
+    e.stopPropagation();
+  }
+
   renderChart() {
     try {
       const mermaidDiv = document.querySelector(".mermaid");
       if (mermaidDiv) {
-        mermaidDiv.innerHTML = ""; // Clear the existing content
-        mermaidDiv.innerHTML = this.props.chartmd; // Set new markdown content
+        // make sure to use textContent and not innerHtml, as our markdown can contain html
+        mermaidDiv.textContent = ""; // Clear the existing content
+        mermaidDiv.textContent = this.props.chartmd; // Set new markdown content
         mermaid.contentLoaded(); // Inform Mermaid to process the new content
+
+        var collapseButtons = document.querySelectorAll(".collapse-expand");
```

Comment:
```suggestion
        const collapseButtons = document.querySelectorAll(".collapse-expand");
```
Safer, since this isn't reassigned.

Reply:
Fixed.